File: Data/Examples/Android/McAndroidTester/app/src/main/java/com/elbit/mapcore/mcandroidtester/managers/EditmodeCallbackManager.java
package com.elbit.mapcore.mcandroidtester.managers;


import java.util.ArrayList;
import java.util.List;

import com.elbit.mapcore.Interfaces.General.IMcEditMode;
import com.elbit.mapcore.Interfaces.Map.IMcMapViewport;
import com.elbit.mapcore.Interfaces.OverlayManager.IMcObject;
import com.elbit.mapcore.Interfaces.OverlayManager.IMcObjectSchemeItem;
import com.elbit.mapcore.Structs.SMcVector3D;

/**
 * Created by tc97803 on 28/09/2016.
 */
public class EditmodeCallbackManager implements IMcEditMode.ICallback {

    private List<IMcEditMode.ICallback> m_registeredComponents;

    private IMcEditMode m_editMode;

    public EditmodeCallbackManager(IMcEditMode editMode)
    {
        this.m_editMode = editMode;
        try {
            m_editMode.SetEventsCallback(this);
        } catch (Exception e) {
            e.printStackTrace();
        }

        m_registeredComponents = new ArrayList<IMcEditMode.ICallback>();
    }

    public void SetEventsCallback(IMcEditMode.ICallback callback)
    {
        if (!m_registeredComponents.contains(callback))
            m_registeredComponents.add(callback);
    }

    public void UnregisterEventsCallback(IMcEditMode.ICallback callback)
    {
        if (callback == null)
            m_registeredComponents.clear();
        else
        {
            if (m_registeredComponents.contains(callback))
                m_registeredComponents.remove(callback);
        }
    }


    @Override
    public void ExitAction(int nExitCode) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.ExitAction(nExitCode);
        }
    }

    @Override
    public void NewVertex(IMcObject iMcObject, IMcObjectSchemeItem iMcObjectSchemeItem, SMcVector3D sMcVector3D, SMcVector3D sMcVector3D1, int i, double v) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.NewVertex(iMcObject, iMcObjectSchemeItem, sMcVector3D, sMcVector3D1, i, v);
        }
    }

    @Override
    public void PointDeleted(IMcObject iMcObject, IMcObjectSchemeItem iMcObjectSchemeItem, SMcVector3D sMcVector3D, SMcVector3D sMcVector3D1, int i) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.PointDeleted(iMcObject, iMcObjectSchemeItem, sMcVector3D, sMcVector3D1, i);
        }
    }

    @Override
    public void PointNewPos(IMcObject iMcObject, IMcObjectSchemeItem iMcObjectSchemeItem, SMcVector3D sMcVector3D, SMcVector3D sMcVector3D1, int i, double v, boolean b) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.PointNewPos(iMcObject,  iMcObjectSchemeItem,  sMcVector3D,  sMcVector3D1,  i,  v,  b);
        }
    }

    @Override
    public void ActiveIconChanged(IMcObject iMcObject, IMcObjectSchemeItem iMcObjectSchemeItem, IMcEditMode.EPermission ePermission, int i) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.ActiveIconChanged(iMcObject,  iMcObjectSchemeItem,  ePermission,  i);
        }
    }

    @Override
    public void InitItemResults(IMcObject iMcObject, IMcObjectSchemeItem iMcObjectSchemeItem, int i) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.InitItemResults(iMcObject, iMcObjectSchemeItem, i);
        }
    }

    @Override
    public void EditItemResults(IMcObject iMcObject, IMcObjectSchemeItem iMcObjectSchemeItem, int i) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.EditItemResults(iMcObject, iMcObjectSchemeItem, i);
        }
    }

    @Override
    public void DragMapResults(IMcMapViewport iMcMapViewport, SMcVector3D sMcVector3D) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.DragMapResults(iMcMapViewport, sMcVector3D);
        }
    }

    @Override
    public void RotateMapResults(IMcMapViewport iMcMapViewport, float v, float v1) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.RotateMapResults(iMcMapViewport, v, v1);
        }
    }

    @Override
    public void DistanceDirectionMeasureResults(IMcMapViewport iMcMapViewport, SMcVector3D sMcVector3D, SMcVector3D sMcVector3D1, double v, double v1) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.DistanceDirectionMeasureResults(iMcMapViewport, sMcVector3D, sMcVector3D1, v, v1);
        }
    }

    @Override
    public void DynamicZoomResults(IMcMapViewport iMcMapViewport, float v, SMcVector3D sMcVector3D) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.DynamicZoomResults(iMcMapViewport, v, sMcVector3D);
        }
    }

    @Override
    public void CalculateHeightResults(IMcMapViewport iMcMapViewport, double v, SMcVector3D[] sMcVector3Ds, int i) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.CalculateHeightResults(iMcMapViewport, v, sMcVector3Ds, i);
        }
    }

    @Override
    public void CalculateVolumeResults(IMcMapViewport iMcMapViewport, double v, SMcVector3D[] sMcVector3Ds, int i) {
        for (IMcEditMode.ICallback callback : m_registeredComponents)
        {
            callback.CalculateVolumeResults(iMcMapViewport, v, sMcVector3Ds, i);
        }
    }


}

File: Data/Examples/CSharpTester/MCTester/GUI/Managers/ObjectWorld/EditmodeCallbackManager.cs
using System;
using System.Collections.Generic;
using System.Text;
using MapCore;
using UnmanagedWrapper;

namespace MCTester.Managers.ObjectWorld
{
    public delegate void ExitActionEventArgs(int nExitCode);
    public delegate void NewVertexEventArgs(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex, double dAngle);
    public delegate void PointDeletedEventArgs(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex);
    public delegate void PointNewPosEventArgs(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex, double dAngle, bool bDownOnHeadPoint);
    public delegate void ActiveIconChangedEventArgs(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNEPermission eIconPermission, uint uIconIndex);
    public delegate void InitItemResultsEventArgs(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, int nExitCode);
    public delegate void EditItemResultsEventArgs(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, int nExitCode);
    public delegate void DragMapResultsEventArgs(IDNMcMapViewport pViewport, DNSMcVector3D NewCenter);
    public delegate void RotateMapResultsEventArgs(IDNMcMapViewport pViewport, float fNewYaw, float fNewPitch);
    public delegate void DistanceDirectionMeasureResultsEventArgs(IDNMcMapViewport pViewport, DNSMcVector3D WorldVertex1, DNSMcVector3D WorldVertex2, double dDistance, double dAngle);
    public delegate void DynamicZoomResultsEventArgs(IDNMcMapViewport pViewport, float fNewScale, DNSMcVector3D NewCenter);
    public delegate void CalculateHeightResultsEventArgs(IDNMcMapViewport pViewport, double dHeight, DNSMcVector3D[] coords, int status);
    public delegate void CalculateVolumeResultsEventArgs(IDNMcMapViewport pViewport, double dVolume, DNSMcVector3D[] coords, int status);

    public class EditmodeCallbackManager : IDNEditModeCallback
    {
        public event ExitActionEventArgs ExitActionEvent;
        public event NewVertexEventArgs NewVertexEvent;
        public event PointDeletedEventArgs PointDeletedEvent;
        public event PointNewPosEventArgs PointNewPosEvent;
        public event ActiveIconChangedEventArgs ActiveIconChangedEvent;
        public event InitItemResultsEventArgs InitItemResultsEvent;
        public event EditItemResultsEventArgs EditItemResultsEvent;
        public event DragMapResultsEventArgs DragMapResultsEvent;
        public event RotateMapResultsEventArgs RotateMapResultsEvent;
        public event DistanceDirectionMeasureResultsEventArgs DistanceDirectionMeasureResultsEvent;
        public event DynamicZoomResultsEventArgs DynamicZoomResultsEvent;
        public event CalculateHeightResultsEventArgs CalculateHeightResultsEvent;
        public event CalculateVolumeResultsEventArgs CalculateVolumeResultsEvent;

        private List<IDNEditModeCallback> m_registeredComponents;

        private IDNMcEditMode m_editMode;

        public EditmodeCallbackManager(IDNMcEditMode editMode)
        {
            this.m_editMode = editMode;
            m_editMode.SetEventsCallback(this);

            m_registeredComponents = new List<IDNEditModeCallback>();
        }

        public void SetEventsCallback(IDNEditModeCallback callback)
        {
            if (!m_registeredComponents.Contains(callback))
                m_registeredComponents.Add(callback);
        }

        public void UnregisterEventsCallback(IDNEditModeCallback callback)
        {
            if (callback == null)
                m_registeredComponents.Clear();
            else
            {
                if (m_registeredComponents.Contains(callback))
                    m_registeredComponents.Remove(callback);
            }                
        }

        #region IDNEditModeCallback Members

        public void ExitAction(int nExitCode)
        {
            if (ExitActionEvent!=null)
            {
                ExitActionEvent(nExitCode);
            }
            
            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.ExitAction(nExitCode);
            }
        }

        public void NewVertex(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex, double dAngle)
        {
            if (NewVertexEvent!=null)
            {
                NewVertexEvent(pObject, pItem, WorldVertex, ScreenVertex, uVertexIndex, dAngle);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.NewVertex(pObject, pItem, WorldVertex, ScreenVertex, uVertexIndex, dAngle);
            }
        }

        public void PointDeleted(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex)
        {
            if (PointDeletedEvent!=null)
            {
                PointDeletedEvent(pObject, pItem, WorldVertex, ScreenVertex, uVertexIndex);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.PointDeleted(pObject, pItem, WorldVertex, ScreenVertex, uVertexIndex);
            }
        }

        public void PointNewPos(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex, double dAngle, bool bDownOnHeadPoint)
        {
            if (PointNewPosEvent != null)
            {
                PointNewPosEvent(pObject, pItem, WorldVertex, ScreenVertex, uVertexIndex, dAngle, bDownOnHeadPoint);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.PointNewPos(pObject, pItem, WorldVertex, ScreenVertex, uVertexIndex, dAngle, bDownOnHeadPoint);
            }
        }

        public void ActiveIconChanged(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNEPermission eIconPermission, uint uIconIndex)
        {
            if (ActiveIconChangedEvent != null)
            {
                ActiveIconChangedEvent(pObject, pItem, eIconPermission, uIconIndex);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.ActiveIconChanged(pObject, pItem, eIconPermission, uIconIndex);
            }
        }

        public void InitItemResults(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, int nExitCode)
        {
            if (InitItemResultsEvent!=null)
            {
                InitItemResultsEvent(pObject, pItem, nExitCode);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.InitItemResults(pObject, pItem, nExitCode);
            }
        }

        public void EditItemResults(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, int nExitCode)
        {
            if (EditItemResultsEvent != null)
            {
                EditItemResultsEvent(pObject, pItem, nExitCode);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.EditItemResults(pObject, pItem, nExitCode);
            }
        }

		public void DragMapResults(IDNMcMapViewport pViewport, DNSMcVector3D NewCenter)
        {
            if (DragMapResultsEvent != null)
            {
                DragMapResultsEvent(pViewport, NewCenter);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.DragMapResults(pViewport, NewCenter);
            }
        }

		public void RotateMapResults(IDNMcMapViewport pViewport, float fNewYaw, float fNewPitch)
        {
            if (RotateMapResultsEvent != null)
            {
                RotateMapResultsEvent(pViewport, fNewYaw, fNewPitch);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.RotateMapResults(pViewport, fNewYaw, fNewPitch);
            }
        }

		public void DistanceDirectionMeasureResults(IDNMcMapViewport pViewport, DNSMcVector3D WorldVertex1, DNSMcVector3D WorldVertex2, double dDistance, double dAngle)
        {
            if (DistanceDirectionMeasureResultsEvent != null)
            {
                DistanceDirectionMeasureResultsEvent(pViewport, WorldVertex1, WorldVertex2, dDistance, dAngle);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.DistanceDirectionMeasureResults(pViewport, WorldVertex1, WorldVertex2, dDistance, dAngle);
            }
        }

		public void DynamicZoomResults(IDNMcMapViewport pViewport, float fNewScale, DNSMcVector3D NewCenter)
        {
            if (DynamicZoomResultsEvent != null)
            {
                DynamicZoomResultsEvent(pViewport, fNewScale, NewCenter);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.DynamicZoomResults(pViewport, fNewScale, NewCenter);
            }
        }

		public void CalculateHeightResults(IDNMcMapViewport pViewport, double dHeight, DNSMcVector3D[] coords, int status)
        {
            if (CalculateHeightResultsEvent != null)
            {
                CalculateHeightResultsEvent(pViewport, dHeight, coords, status);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.CalculateHeightResults(pViewport, dHeight, coords, status);
            }
        }

		public void CalculateVolumeResults(IDNMcMapViewport pViewport, double dVolume, DNSMcVector3D[] coords, int status)
        {
            if (CalculateVolumeResultsEvent != null)
            {
                CalculateVolumeResultsEvent(pViewport, dVolume, coords, status);
            }

            foreach (IDNEditModeCallback callback in m_registeredComponents)
            {
                callback.CalculateVolumeResults(pViewport, dVolume, coords, status);
            }
        }

        #endregion
    }
}
